deepinv/physics/mri.py

```diff
@@ -1,38 +1,104 @@
+        return torch.view_as_complex(x.moveaxis(1, -1).contiguous())
+
+    @staticmethod
+    def from_torch_complex(x: Tensor):
+        """[B,...,H,W] complex -> [B,2,...,H,W] real"""
+        return torch.view_as_real(x).moveaxis(-1, 1)
+
+    @staticmethod
+    def ifft(x: Tensor, dim=(-2, -1), norm="ortho"):
+        """Centered, orthogonal ifft
+
+        :param torch.Tensor x: input kspace of complex dtype of shape [B,...] where ... is all dims to be transformed
+        :param tuple dim: fft transform dims, defaults to (-2, -1)
+        :param str norm: fft norm, see docs for :meth:`torch.fft.fftn`, defaults to "ortho"
+        """
+        x = torch.fft.ifftshift(x, dim=dim)
+        x = torch.fft.ifftn(x, dim=dim, norm=norm)
+        return torch.fft.fftshift(x, dim=dim)
+
+    @staticmethod
+    def fft(x: Tensor, dim=(-2, -1), norm="ortho"):
+        """Centered, orthogonal fft
+
+        :param torch.Tensor x: input image of complex dtype of shape [B,...] where ... is all dims to be transformed
+        :param tuple dim: fft transform dims, defaults to (-2, -1)
+        :param str norm: fft norm, see docs for :meth:`torch.fft.fftn`, defaults to "ortho"
+        """
+        x = torch.fft.ifftshift(x, dim=dim)
+        x = torch.fft.fftn(x, dim=dim, norm=norm)
+        return torch.fft.fftshift(x, dim=dim)
+
+
+class MRI(MRIMixin, DecomposablePhysics):
+    r"""
+    Single-coil accelerated 2D or 3D magnetic resonance imaging.
+
+    The linear operator operates in 2D slices or 3D volumes and is defined as
 
     .. math::
 
-        y = SFx
+        y = MFx
 
-    where :math:`S` applies a mask (subsampling operator), and :math:`F` is the 2D discrete Fourier Transform.
+    where :math:`M` applies a mask (subsampling operator), and :math:`F` is the 2D or 3D discrete Fourier Transform.
     This operator has a simple singular value decomposition, so it inherits the structure of
     :meth:`deepinv.physics.DecomposablePhysics` and thus have a fast pseudo-inverse and prox operators.
 
-    The complex images :math:`x` and measurements :math:`y` should be of size (B, 2, H, W) where the first channel corresponds to the real part
-    and the second channel corresponds to the imaginary part.
+    The complex images :math:`x` and measurements :math:`y` should be of size (B, C,..., H, W) with C=2, where the first channel corresponds to the real part
+    and the second channel corresponds to the imaginary part. The ``...`` is an optional depth dimension for 3D MRI data.
 
     A fixed mask can be set at initialisation, or a new mask can be set either at forward (using ``physics(x, mask=mask)``) or using ``update_parameters``.
 
     .. note::
 
         We provide various random mask generators (e.g. Cartesian undersampling) that can be used directly with this physics. See e.g. :class:`deepinv.physics.generator.mri.RandomMaskGenerator`
+        If mask is not passed, a blank one is used.
```

Comment:
the word blank makes me think of an empty mask, should we say "a mask full of ones is passed (no acceleration)"?

Reply:
done